changefeedccl: TestChangefeedEndTimeWithCursor/kafka fails in CI

https://teamcity.cockroachdb.com/viewLog.html?buildId=11728871&buildTypeId=Cockroach_BazelEssentialCi

```
=== RUN   TestChangefeedEndTimeWithCursor/kafka
    helpers_test.go:856: making server as secondary tenant
    helpers_test.go:934: making kafka feed factory
    changefeed_test.go:7189: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:7189
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1090
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1125
        	Error:      	Received unexpected error:
        	            	still waiting for job status; current running
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:434
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:81
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoonError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:51
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:425
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedEndTimeWithCursor.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:7189
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1090
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1125
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1576
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1598
        	            	Wraps: (2) still waiting for job status; current running
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestChangefeedEndTimeWithCursor/kafka
=== NAME  TestChangefeedEndTimeWithCursor
    changefeed_test.go:7206: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0fb356fafa2ae761a107637f3a1fdcc7/logTestChangefeedEndTimeWithCursor56448993
    --- FAIL: TestChangefeedEndTimeWithCursor/kafka (51.75s)
```

Jira issue: CRDB-31402

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -7185,7 +7185,11 @@ func TestChangefeedEndTimeWithCursor(t *testing.T) {
 		defer closeFeed(t, feed)
 
 		// Don't care much about the values emitted (tested elsewhere) -- all
-		// we want to make sure is that the feed terminates.
+		// we want to make sure is that the feed terminates.  However, we do need
+		// to consume those values since some of the test sink implementations (kafka)
+		// will block.
+		defer DiscardMessages(feed)()
+
 		testFeed := feed.(cdctest.EnterpriseTestFeed)
 		require.NoError(t, testFeed.WaitForStatus(func(s jobs.Status) bool {
 			return s == jobs.StatusSucceeded

--- a/pkg/ccl/changefeedccl/testfeed_test.go
+++ b/pkg/ccl/changefeedccl/testfeed_test.go
@@ -1928,6 +1928,35 @@ func (k *kafkaFeed) Close() error {
 	return errors.CombineErrors(k.jobFeed.Close(), k.tg.wait())
 }
 
+// DiscardMessages spins up a goroutine to discard messages produced into the
+// sink. Returns cleanup function.
+func DiscardMessages(f cdctest.TestFeed) func() {
+	switch t := f.(type) {
+	case *kafkaFeed:
+		ctx, cancel := context.WithCancel(context.Background())
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-t.shutdown:
+					return
+				case <-t.source:
+				}
+			}
+		}()
+		return func() {
+			cancel()
+			wg.Wait()
+		}
+	default:
+		return func() {}
+	}
+}
+
 type webhookFeedFactory struct {
 	enterpriseFeedFactory
 	useSecureServer bool

